Fix Windows compatibility issues and AGE support

SlugFiller opened the pull request:
Builds on top of #362. Fixed parts that weren't working correctly, ensured it builds on Python 3.11 without errors, fixed issues with AGE, and added Windows-specific documentation.

Only thing missing is GPG support. However, I couldn't find a user-friendly non-CygWin GPG client for Windows. And I've never used GPG in the past. This makes it difficult to test, debug, or even just document GPG usage.

All other tools are confirmed to work, and appropriate Windows-specific tooling for them has been documented.

SlugFiller commented:
Should I add a Windows build GitHub Workflow to this? I'm not too familiar with Workflows, but it shouldn't be too difficult. The question is whether CI for Windows is a priority, and if it should be part of this PR or separate.

romanz commented:
Many thanks for the contribution!
Will review this weekend :)

romanz reviewed (commented):
Initial review - please take a look below:

romanz commented:
Also the CI seems to fail - please take a look:
https://github.com/romanz/trezor-agent/actions/runs/6114735009/job/16605844896?pr=436#step:5:74

SlugFiller commented on setup.py:
```diff
@@ -31,8 +31,9 @@
         'pymsgbox>=1.0.6',
         'semver>=2.2',
         'unidecode>=0.4.20',
+        'pypiwin32;sys_platform=="win32"'
```
You are correct. `pypiwin32` is an old package since before `pywin32` has been available on `pip`. It's been unnecessary since around 2018.

SlugFiller commented on libagent/ssh/protocol.py:
```diff
@@ -53,7 +53,7 @@ def msg_code(name):
     """Convert integer message code into a string name."""
     ids = {v: k for k, v in COMMANDS.items()}
-    return ids[code]
+    return ids[code] if code in ids else str(code)
```
It's a sanity check. `msg_name` is specifically used in the case the method is unknown. In the case of a malformed packet, or a future extension to the protocol, this is the difference between closing the connection with an unsupported method error, and crashing the agent itself.

romanz commented:
Please fix the code style issues: 
```
libagent/win_server.py:20:1: E302 expected 2 blank lines, found 1
libagent/win_server.py:42:18: E261 at least two spaces before inline comment
libagent/win_server.py:47:1: W293 blank line contains whitespace
libagent/win_server.py:55:101: E501 line too long (107 > 100 characters)
libagent/win_server.py:60:9: E722 do not use bare 'except'
libagent/win_server.py:119:101: E501 line too long (103 > 100 characters)
libagent/win_server.py:121:101: E501 line too long (106 > 100 characters)
libagent/win_server.py:194:9: E722 do not use bare 'except'
libagent/ssh/__init__.py:36:1: E302 expected 2 blank lines, found 1
libagent/ssh/__init__.py:206:35: E261 at least two spaces before inline comment
libagent/ssh/__init__.py:208:37: E261 at least two spaces before inline comment
libagent/ssh/__init__.py:315:22: W605 invalid escape sequence '\p'
```
https://github.com/romanz/trezor-agent/actions/runs/6123995379/job/16624649832?pr=436

romanz commented:
@Pandapip1 could you please help testing this PR?

Pandapip1 commented:
Sure! Are there any build instructions I can follow?

For now, I'm making a program using USB/IP that mostly implements the smart card USB interface for use with the Trezor. But if you could get this working, that'd be better!

SlugFiller commented:
* Made my best effort to lint
* Fixed GPG. Tested by signing this commit (May upload my public key to GitHub later)
* Refactored win_server so it now has the same interface as a regular UNIX socket, and doesn't require a separate server loop
* Device UI will now attempt to get `pinentry` location from `gpg-conf --list-components` instead of assuming it is in the path. It will fall back to just `pinentry` if it fails.

Currently not working (On Windows):
* Creating identity as subkey. Since it requires making an outgoing communication to `gpg-agent`, which is a slightly different beast from receiving an incoming one.

SlugFiller commented:
Improved Ctrl+C behavior. Before, running the GPG agent directly with no client would cause it to freeze, and be uninterruptible. Also improved behavior of named pipe so it remains open while waiting for a connection, and closes instantly on Ctrl+C.

SlugFiller commented:
Fixed, but didn't test, GPG subkey registration.

SlugFiller commented:
More linting

SlugFiller commented:
isorted

romanz commented on libagent/gpg/keyring.py:
```diff
@@ -29,19 +34,23 @@ def get_agent_sock_path(env=None, sp=subprocess):
-    lines = output.strip().split(b'\n')
-    dirs = dict(line.split(b':', 1) for line in lines)
+    lines = re.split('[\n\r]+', output.decode('utf-8'))
+    dirs = {k: urllib.parse.unquote(v).encode() for k, v in (
```
Why `urllib.parse.unquote` is needed here (and below)?

SlugFiller replied:
Because the components are URL-encoded. Notably, paths which contain `:`, e.g. `C:\\Program Files (x86)\\GnuPG\\bin\\gpg.exe` are encoded as `C%3a\\Program Files (x86)\\GnuPG\\bin\\gpg.exe`

romanz commented on libagent/util.py:
```diff
@@ -258,6 +259,30 @@ def assuan_serialize(data):
     return data
 
 
+def escape_cmd_quotes(str):
```
Consider using https://docs.python.org/3/library/stdtypes.html#str.translate

SlugFiller replied:
Good point. Done.

romanz commented on libagent/win_server.py:
```diff
@@ -0,0 +1,357 @@
+                            raise OSError('Could not read port for socket {0}'.format(pipe_name))
+                        port.write(c)
+                    port = int(port).getvalue()
+                    key_len = 0
```
Where is `key_len` updated?

romanz commented on libagent/win_server.py:
```diff
@@ -0,0 +1,357 @@
+                    key_len = 0
+                    key = io.BytesIO()
+                    while key:
+                        c = f.read(16-keylen)
```
Did you mean `key_len`?

SlugFiller replied:
Lint already caught this. Fixed.

romanz commented on libagent/win_server.py:
```diff
@@ -0,0 +1,357 @@
+            return (pipe, self.pipe_name)
+
+
+class Client:
```
Please explain the logic behind using both a socket and a named pipe here.

SlugFiller replied:
Tool dependent. Windows doesn't have AF_UNIX sockets (well, it does since Windows 10, but it is still rarely used), so different services ported from Linux circumvent this issue in different ways. OpenSSH uses named pipes. GPG uses a localhost socket, locked by a nonce that is written to a file. `win_server` automatically switches between the two based on whether the socket path is a string (coming from an argument like `--sock-path) or byte string (return value from `gpg.keyring.get_agent_sock_path`)

A more robust way may be to check for `\\.\pipe\` prefix. But given that the addresses already come in two different types, they would have to be normalized first.

SlugFiller commented:
More lints, and minor fixes. Hopefully CI finally passes this time.

SlugFiller commented:
Question. Since rebasing the previous PR killed its signature, should I try rebasing so my commit looks like a merge?

SlugFiller commented:
Another question. Why does `keyring.get_agent_sock_path` call `gpgconf --list-dirs`, parses all the directories, and then extracts only `agent-socket` instead of just calling `gpgconf --list-dirs agent-socket`?

SlugFiller commented:
Okay, after playing around with GPG for a day, I made some minor changes, and updated the docs a lot. And also, GPG is awesome, and I don't know why people are trying to find replacements for it.

I also have a pretty good idea of how to implement an `add` command that will allow creating an additional hardware-backed identity on the same GPG profile. I feel like such a feature is necessary, since you don't want to use an identity with an `@users.noreply.github.com` email outside of GitHub, yet GitHub won't allow you to verify with an identity that doesn't have a matching email.

But adding it to this PR would be too many features in one PR. So, as soon as all the fixes are done and this PR is merged, I'll work on that feature.

romanz commented:
> Since rebasing the previous PR killed its signature, should I try rebasing so my commit looks like a merge?

It's OK to rebase without signing the commit, if you prefer.

romanz commented:
> Why does `keyring.get_agent_sock_path` call `gpgconf --list-dirs`, parses all the directories, and then extracts only `agent-socket` instead of just calling `gpgconf --list-dirs agent-socket`?

Good catch - if `gpgconf --list-dirs agent-socket` works, let's use it :)

romanz commented on libagent/gpg/__init__.py:
```diff
@@ -135,32 +141,44 @@ def run_init(device_type, args):
-    check_call(['mkdir', '-p', homedir])
-    check_call(['chmod', '700', homedir])
+    # Prepare the key before making any changes
+    key_id, public_key = export_public_key(device_type, args)
```
nit:
```suggestion
    key_id, public_key_bytes = export_public_key(device_type, args)
```

romanz commented:
> Sure! Are there any build instructions I can follow?

@Pandapip1 Please take a look at https://github.com/romanz/trezor-agent/blob/d0d0c83638f29d631a0e56f6e15e87d30fe9ea9a/doc/README-Windows.md

romanz commented:
LGTM (reviewed most of the code - except `win_server.py`), many thanks!

SlugFiller commented on libagent/ssh/__init__.py:
```diff
@@ -241,13 +275,16 @@ def _dummy_context():
+        log.error('running in foreground mode requires specifying %s', SOCK_TYPE_PATH)
+        sys.exit(1)
+    elif sys.platform == 'win32':
+        suffix = random.choices(string.ascii_letters, k=10)
```
`mkstemp` is not applicable because it will try to create and open the file, which can't be done for a pipe address. `mktemp` will also try to check for the file's existence, which may or may not work for a pipe address, and is deprecated.

SlugFiller commented:
> It's OK to rebase without signing the commit, if you prefer.

Already rebased as a merge. Was fairly simple to do

> Good catch - if `gpgconf --list-dirs agent-socket` works, let's use it :)

It does, used.

Fixed pydocs. According to `tox.ini` this should be the last step. (Unless coverage has an issue with me too)

SlugFiller commented:
Okay, *now* it should lint. **Crosses fingers**

SlugFiller commented:
Drat. Seems like the `agent_sock_path` optimization either only works on Windows, or only works with a specific version of GPG. Reverted that change.

SlugFiller commented:
No wait. The issue is with the test. It uses a mock subprocess

SlugFiller commented:
Okay, how do I fix it? If I change the mock process, it's going to be a trivial test. If I use a real call, would GPG be available in the testing environment? Alternately, I can just remove the test.

SlugFiller commented:
Okay, for now I'm attempting with a real process. If it goes well, great. If not, I'm going to need help with this.

SlugFiller commented:
Okay, that went better than I was expecting. Corrected the test. Should work now.

SlugFiller commented:
Optimized the GPG init so it doesn't create files that don't need to stay there.

SlugFiller commented:
Not meaning to rush anyone, but CI has finally passed, I've applied the suggested fixes (where applicable), and while there's a small mountain of fixes I want to add, none of them are specific to Windows, i.e. they should not be part of this PR. So, what is the next step? Waiting for testers? Reviewing the last few bits of code added?

romanz reviewed (approved):
No problem, and many thanks for the work!

gtbuchanan commented:
@SlugFiller Thank you for running with this and actually preserving my contribution. I was obviously out of my comfort zone. GPG support is coming too? Nice work!

SlugFiller commented:
@gtbuchanan While I ended up changing a lot, your contribution was a huge factor in being able to make this PR, or at the very least, make it this quickly (Less than a week from conception to merge). At a bare minimum, having a working named pipes with overlapped io implementation in pywin32 I could reference, got the hard part out of the way.
